kolibri/core/assets/src/components.js

```diff
@@ -149,6 +73,132 @@ var CrudCollection = Mn.CollectionView.extend({
+    // In other words, if the view instance is given a model attribute when instantiated, then the argument
+    // to this function will be a copy of that model's .attributes hash.
+    template: function() {
+        var html =
+            '{{#unless valid}}<span>Invalid Input!</span><br />{{/unless}}' +
+            '{{#each create}}' +
+                '<input data-attr="{{ name }}" type="text" placeholder="{{ name }}" value="{{ default }}" /><br />' +
+            '{{/each}}' +
+            '<button class="create flat-button">Create</button>';
+        return Handlebars.compile(html)({create: this.create, valid: this.valid});
+    },
+
+    triggers: {
+        // Re-triggers the specified DOM event (from its template) as a Backbone.Event on this view.
+        'click .create': 'create'
+    },
+
+    initialize: function(options) {
+        // Some options when passed to the constructor are automatically bound to the view instance.
+        // This is the case for the "model" attribute, so passing a model to a view constructor means that the
+        // instance has it available automatically as "this.model".
+        // Custom options (like "create" below) must be bound to the view instance manually, or otherwise handled
+        // by the initialize function.
+        this.create = _.map(options.create, function(attr) {
+            return {
+                name: attr,
+                default: ''
+            };
+        });
+        this.validators = options.validators;
+        this.valid = true;
+        // Since "template" is a callback, bindAll ensures "this" refers the view instance.
+        _.bindAll(this, 'template');
+    },
+
+    // For event "xyz", you can specify a handler without having to set up a listener using the camel-cased
+    // "onXyz" convention.
+    // E.g. "create" event is handled automatically by the "onCreate" function if it exists.
+    onCreate: function() {
+        if ( this.validate() ) {
+            _.forEach(this.create, _.bind(function (obj) {
+                var attr = obj.name;
+                this.model.set(attr, this.getAttrVal(attr));
+            }, this));
+            this.collection.add(this.model);
+            this.trigger('closeModal');
+        } else {
+            // Shows an "Invalid!" message.
+            this.rerender = true;
+            this.render();
+        }
+    },
+
+    onBeforeRender: function() {
+        // Before re-rendering, make sure that values user has already entered aren't trashed
+        if (this.rerender) {
+            var self = this;
+            this.create = _.map(this.create, function(old) {
+                return {
+                    name: old.name,
+                    default: self.getAttrVal(old.name)
+                };
+            });
+        }
+    },
+
+    getAttrVal: function(attr) {
+        var el = this.$el.find('input[data-attr=' + attr + ']');
+        return $(el).val();
+    },
```

Comment:
I see what you mean -- IMO the data attribute stuff is kind of unavoidable. I suspect it's the kind of thing going on behind the scenes for forms in any framework, but you're right that here it's our responsibility to manage the complexity. 

As for onBeforeRender, that's necessitated to keep the form "nice" on rerender, kind of in line with what you had mentioned about input elements losing focus on re-render. In this case, it's so that intermediary values are kept even when validation fails.

The trade-off is whether the global app state should granularly manage the state of each form (i.e. have a disabled attribute for each field, a focus attribute, an attribute representing the value of each input element) vs hiding many of the details by managing some state in the form view itself. Since they're highly interactive, forms seem to experience a "state explosion". Moreover any element that deals with user input I think will face similar issues in terms of reacting to user's actions -- should user input be modeled as a change of global state, or do we introduce local state with the aim of simplifying global state?